### PR TITLE
[4.3] Several fixes to single tag view from com_tags

### PR DESCRIPTION
**Pull request by Hackwar**
### Summary of Changes
This is the result of a codereview of the tag view from com_tags. There are several changes in here, which fix broken behavior in this one view and clean up code. I'm doing my best to document everything in here... This is a complex change, but unfortunately I don't see how to simplify this and/or split this up into several PRs. For all existing sites, this PR will not result in new errors, but rather fix existing issues.

Let's start into the changes in here:
- The view class has been reordered to mimic the article view of com_content. This means, that data is assigned directly to a class attribute and not to a temporary variable, after the data is collected from the model, the parameters are merged appropriately and then the plugin events are run.
- A bunch of docblocks have been fixed.
- The feature to count hits on tags has been moved from the view to the controller, similar to how it is done in com_content. An option to disable this has been added as well.
- `$this->item` in the view class contains an array of tags and up till now only the first tag was checked against the menu item, etc. Now all the tags are checked.
- At several places, the view class checked the access levels. This is something that we are doing in the model (already) not in the view.
- Merging of parameters was unnecessarily complex.
- In the tag model, the `$item` attribute wasn't defined. `$tag` at the same time wasn't used. With `$item` being initialised as an array, we can remove a bunch of checks. At the same time we can now properly throw a 404 if the number of tag IDs doesn't match the number of tags found in the database matching these IDs. (we might want to rewrite the `TagModel::getItem()` method to use a single query instead of loading each tag separately with a Table call...)

### Testing Instructions
Basically there should be no difference on an existing site between the current code and after applying this PR.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [X] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [X] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: administrator/components/com_tags/config.xml [hunk 1]
```diff
@@ -211,6 +211,17 @@
 			<option value="current_language">JCURRENT</option>
 		</field>
 
+		<field
+			name="record_hits"
+			type="radio"
+			label="JGLOBAL_RECORD_HITS_LABEL"
+			layout="joomla.form.field.radio.switcher"
+			default="1"
+		>
```
Comment (brianteeman) [inline administrator/components/com_tags/config.xml]:
```suggestion
			>
```

File: components/com_tags/src/View/Tag/HtmlView.php [hunk 2]
```diff
@@ -10,6 +10,8 @@
 
 namespace Joomla\Component\Tags\Site\View\Tag;
 
+use Joomla\CMS\Menu\MenuItem;
+use Joomla\CMS\Object\CMSObject;
 use Joomla\CMS\Factory;
 use Joomla\CMS\MVC\View\GenericDataException;
 use Joomla\CMS\MVC\View\HtmlView as BaseHtmlView;
```
Comment (brianteeman) [inline components/com_tags/src/View/Tag/HtmlView.php]:
```suggestion
use Joomla\CMS\Factory;
use Joomla\CMS\Menu\MenuItem;
use Joomla\CMS\MVC\View\GenericDataException;
use Joomla\CMS\MVC\View\HtmlView as BaseHtmlView;
use Joomla\CMS\Object\CMSObject;
```

File: components/com_tags/src/View/Tag/HtmlView.php [hunk 3]
```diff
@@ -131,152 +133,105 @@ class HtmlView extends BaseHtmlView
...
-        $children   = $this->get('Children');
-        $parent     = $this->get('Parent');
-        $pagination = $this->get('Pagination');
+        $this->state      = $this->get('State');
+        $this->items      = $this->get('Items');
+        $this->item       = $this->get('Item');
+        $this->children   = $this->get('Children');
+        $this->parent     = $this->get('Parent');
+        $this->pagination = $this->get('Pagination');
+        $this->user   = $this->getCurrentUser();
```
Comment (brianteeman) [inline components/com_tags/src/View/Tag/HtmlView.php]:
```suggestion
        $this->pagination = $this->get('Pagination');
        $this->user       = $this->getCurrentUser();
```

File: components/com_tags/src/Model/TagModel.php [hunk 4]
```diff
@@ -306,10 +302,10 @@ public function getItem($pk = null)
                     return false;
                 }
             }
-        }
 
-        if (!$this->item) {
-            throw new \Exception(Text::_('COM_TAGS_TAG_NOT_FOUND'), 404);
+            if (count($this->item) != count($idsArray)) {
+                throw new \Exception(Text::_('COM_TAGS_TAG_NOT_FOUND'), 404);
+            }
```
Comment (chmst) [inline components/com_tags/src/Model/TagModel.php:308]: This is about items, not about tags. If no items are found for a tag, it is not an exception but quite normal and should be a message.

Comment (Hackwar) [reply components/com_tags/src/Model/TagModel.php:308]: What would you say if I told you, that `$this->items` is the list of content items for a tag combination and `$this->item` is actually an array of tags, which we are trying to display... This is actually correct the way it is and we are checking here if the number of IDs we are supposed to find matches with the number of tags we actually found in our system. Otherwise you could go ahead and create a non-sef URL like `index.php?option=com_tags&view=tag&id[0]=42:TAG-1&id[1]=69:TAG-2` and the system might not contain `TAG-2` and you would only get results for `TAG-1`. It is basically solving the issue we had up till Joomla 4 that you could generate new URLs by simply adding some random garbage at the end and Joomla would only parse it as far as it understood it and ignore the rest, returning duplicate content for this new (garbage) URL.

**Comment by chmst** [comment]
Then I misunderstood that, sorry. I see the problem from your point of view. But as a user  I would be ok to know that there is nothing to see. Without 404 error message.

**Comment by jamfx** [comment]
I have tested this item :white_check_mark: successfully on 80c6c27db3b1b6c91412552a2c9ea6e70ff4b079<br /><br />Testes and no Problems found.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39113">issues.joomla.org/tracker/joomla-cms/39113</a>.</sub>

**Comment by richard67** [comment]
I've restored @jamfx 's test result in the issue tracker since the commits which have invalidated the test count were just clean branch updates.

@Hackwar Please avoid unnecessary branch updates since that invalidates the test counter in the issue tracker and so makes it harder for us to find the PRs which have 2 good tests and causes additional work for restoring the test result. Thanks in advance.

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 47f97582c316e8c8f3d9cc0ad3252036b2868ff7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39113">issues.joomla.org/tracker/joomla-cms/39113</a>.</sub>

**Comment by jwaisner** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39113">issues.joomla.org/tracker/joomla-cms/39113</a>.</sub>

**Comment by cyanama** [comment]
I have tested this item ✅ successfully

Please give us customizable css styles for the image and the title. At the moment you need to customize the raw **img** and **h1** statements with impact for the whole webpage.

thanks
Cyana<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39113">issues.joomla.org/tracker/joomla-cms/39113</a>.</sub>

File: components/com_tags/src/Controller/DisplayController.php [hunk 5]
```diff
@@ -56,6 +57,15 @@ public function display($cachable = false, $urlparams = false)
             'lang'             => 'CMD'
         );
 
+        if ($vName === 'tag') {
```
Comment (laoneo) [inline components/com_tags/src/Controller/DisplayController.php]: This if statement should be written as one, then it is easier to read.

Comment (Hackwar) [reply components/com_tags/src/Controller/DisplayController.php]: Done.

**Comment by richard67** [comment]
RTC is still valid because the change after that was just replacing 3 nested if statements by one with the combined condition, which is ok for me by review.

**Comment by obuisard** [comment]
Thank you Hannes @Hackwar for the PR!
